src/ast/parse.rs

```diff
@@ -0,0 +1,895 @@
+            }
+        }
+        Location { line, col }
+    }
+}
+
+// we do this slightly un-idiomatic thing because `unwrap` and `expect`
+// would print the entire source program without this
+impl Debug for Span {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{}", self)
+    }
+}
+
+impl Display for Span {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let start = self.0.get_location(self.1);
+        let end = self.0.get_location((self.2 - 1).max(self.1));
+        let quote = self.string();
+        match (&self.0.name, start.line == end.line) {
+            (Some(filename), true) => write!(
+                f,
+                "In {}:{}-{} of {filename}: {quote}",
+                start.line, start.col, end.col
+            ),
+            (Some(filename), false) => write!(
+                f,
+                "In {}:{}-{}:{} of {filename}: {quote}",
+                start.line, start.col, end.line, end.col
+            ),
+            (None, false) => write!(
+                f,
+                "In {}:{}-{}:{}: {quote}",
+                start.line, start.col, end.line, end.col
+            ),
+            (None, true) => write!(f, "In {}:{}-{}: {quote}", start.line, start.col, end.col),
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+struct Context {
+    source: Arc<SrcFile>,
+    index: usize,
+}
+
+impl Context {
+    fn new(name: Option<String>, contents: &str) -> Context {
+        let mut next = Context {
+            source: Arc::new(SrcFile {
+                name,
+                contents: contents.to_string(),
+            }),
+            index: 0,
+        };
+        next.advance_past_whitespace();
+        next
+    }
+
+    fn current_char(&self) -> Option<char> {
+        self.source.contents[self.index..].chars().next()
+    }
+
+    fn advance_char(&mut self) {
+        assert!(self.index < self.source.contents.len());
+        loop {
+            self.index += 1;
+            if self.source.contents.is_char_boundary(self.index) {
+                break;
+            }
+        }
+    }
+
+    fn advance_past_whitespace(&mut self) {
+        let mut in_comment = false;
+        loop {
+            match self.current_char() {
+                None => break,
+                Some(';') => in_comment = true,
+                Some('\n') => in_comment = false,
+                Some(c) if c.is_whitespace() => {}
+                Some(_) if in_comment => {}
+                Some(_) => break,
+            }
+            self.advance_char();
+        }
+    }
+
+    fn is_at_end(&self) -> bool {
+        self.index == self.source.contents.len()
+    }
+
+    fn span(&self) -> Span {
+        Span(self.source.clone(), self.index, self.index)
+    }
+}
+
+type Res<T> = Result<(T, Span, Context), ParseError>;
+
+trait Parser<T>: Fn(&Context) -> Res<T> + Clone {}
+impl<T, F: Fn(&Context) -> Res<T> + Clone> Parser<T> for F {}
```

Comment:
a thought: add `map` as a method here? Might make some of the stuff a bit more idiomatic rust.

Similar for other combinators that take a parser has their first arg and then some other arg, then return another parser. 

I think this is possible to do right now that RPITIT is in.

Reply:
Turns out that this breaks compile times, see #468. I will fix this.